File: day04/day04.go
package day04

import (
	"fmt"
	"strconv"
	"strings"
)

type secRange struct {
	start int
	end   int
}

type pair [2]secRange

func (p pair) fullyOverlap() bool {
	return p[0].start >= p[1].start && p[0].end <= p[1].end || p[1].start >= p[0].start && p[1].end <= p[0].end
}

func (p pair) overlap() bool {
	return p[0].start >= p[1].start && p[0].start <= p[1].end || p[1].start >= p[0].start && p[1].start <= p[0].end
}

func parseInput(input string) []pair {
	lines := strings.Split(string(input), "\n")
	lines = lines[:len(lines)-1]
	pairs := make([]pair, len(lines))
	for i, line := range lines {
		ranges := strings.Split(line, ",")
		for j, r := range ranges {
			sec := strings.Split(r, "-")
			start, _ := strconv.Atoi(sec[0])
			end, _ := strconv.Atoi(sec[1])
			pairs[i][j] = secRange{start, end}
		}
	}
	return pairs
}

func Solve(input string, test bool, task int) (string, string) {
	pairs := parseInput(input)
	res1, res2 := "", ""
	count := 0

	if task != 2 {
		for _, p := range pairs {
			if p.fullyOverlap() {
				count++
			}
		}
		fmt.Printf("fully overlapping pairs: %v\n", count)
		res1 = strconv.Itoa(count)
	}
	if task != 1 {
		count = 0
		for _, p := range pairs {
			if p.overlap() {
				count++
			}
		}
		fmt.Printf("overlapping pairs: %v\n", count)
		res2 = strconv.Itoa(count)
	}
	return res1, res2
}

File: aoc22_rust/src/day04.rs
use crate::types::Task;

#[derive(Default, Copy, Clone)]
struct SecRange {
    start: i32,
    end: i32,
}

type Pair = [SecRange; 2];

fn fully_overlap(p: &Pair) -> bool {
    p[0].start >= p[1].start && p[0].end <= p[1].end
        || p[1].start >= p[0].start && p[1].end <= p[0].end
}
fn overlap(p: &Pair) -> bool {
    p[0].start >= p[1].start && p[0].start <= p[1].end
        || p[1].start >= p[0].start && p[1].start <= p[0].end
}

fn parse_input(input: &str) -> Vec<Pair> {
    let lines = input.lines();
    let mut pairs: Vec<Pair> = Vec::new();
    for line in lines {
        let ranges = line.split(",");
        let mut pair = [SecRange::default(); 2];
        for (j, r) in ranges.enumerate() {
            let ints: Vec<i32> = r.split("-").map(|e| e.parse::<i32>().unwrap()).collect();
            pair[j] = SecRange {
                start: ints[0],
                end: ints[1],
            };
        }
        pairs.push(pair);
    }
    return pairs;
}

pub fn solve(input: &str, _test: bool, task: Task) -> (String, String) {
    let pairs = parse_input(input);
    let (mut res1, mut res2) = ("".to_string(), "".to_string());

    if !matches!(task, Task::Two) {
        let count = &pairs.iter().filter(|p| fully_overlap(p)).count();
        res1 = count.to_string();
    }
    if !matches!(task, Task::One) {
        let mut count = 0;
        for p in pairs {
            if overlap(&p) {
                count += 1;
            }
        }
        res2 = count.to_string();
    }
    return (res1, res2);
}
